crates/pessimistic-proof-utils/src/local_exit_tree.rs

```diff
@@ -0,0 +1,248 @@
+#[serde_as]
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct LocalExitTreeData<H, const TREE_DEPTH: usize = 32>
+where
+    H: Hasher,
+    H::Digest: Serialize + for<'a> Deserialize<'a>,
+{
+    #[serde_as(as = "[_; TREE_DEPTH]")]
+    layers: [Vec<H::Digest>; TREE_DEPTH],
+    #[serde_as(as = "[_; TREE_DEPTH]")]
+    empty_hash_at_height: [H::Digest; TREE_DEPTH],
+}
+
+#[serde_as]
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct LETMerkleProof<H, const TREE_DEPTH: usize = 32>
+where
+    H: Hasher,
+    H::Digest: Serialize + for<'a> Deserialize<'a>,
+{
+    #[serde_as(as = "[_; TREE_DEPTH]")]
+    siblings: [H::Digest; TREE_DEPTH],
+}
+
+impl<H, const TREE_DEPTH: usize> LocalExitTreeData<H, TREE_DEPTH>
+where
+    H: Hasher,
+    H::Digest: Copy + Default + Serialize + for<'a> Deserialize<'a>,
+{
+    /// Creates a new empty [`LocalExitTreeData`].
+    pub fn new() -> Self {
+        let mut empty_hash_at_height = [H::Digest::default(); TREE_DEPTH];
+        for height in 1..TREE_DEPTH {
+            empty_hash_at_height[height] = H::merge(
+                &empty_hash_at_height[height - 1],
+                &empty_hash_at_height[height - 1],
+            );
+        }
+        LocalExitTreeData {
+            layers: std::array::from_fn(|_| Vec::new()),
+            empty_hash_at_height,
+        }
+    }
+
+    /// Creates a new [`LocalExitTreeData`] and populates its leaves.
+    pub fn from_leaves(leaves: impl Iterator<Item = H::Digest>) -> Self {
+        let mut tree = Self::new();
+
+        for leaf in leaves {
+            tree.add_leaf(leaf);
+        }
+
+        tree
+    }
+
+    /// Appends a leaf to the tree.
+    pub fn add_leaf(&mut self, leaf: H::Digest) -> usize {
+        let leaf_index = self.layers[0].len();
+        assert_eq!(leaf_index >> TREE_DEPTH, 0, "Too many leaves.");
+        self.layers[0].push(leaf);
+        let mut index = leaf_index;
+        let mut entry = leaf;
+        for height in 0..TREE_DEPTH - 1 {
+            let sibling = self.layers[height]
+                .get(index ^ 1)
+                .unwrap_or(&self.empty_hash_at_height[height]);
+            entry = if index & 1 == 1 {
+                H::merge(&sibling, &entry)
+            } else {
+                H::merge(&entry, sibling)
+            };
+            index >>= 1;
+            if index < self.layers[height + 1].len() {
+                self.layers[height + 1][index] = entry;
+            } else {
+                self.layers[height + 1].push(entry);
+            }
+        }
+
+        leaf_index
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.layers[0].is_empty()
+    }
+
+    /// Returns the root of the tree.
+    pub fn get_root(&self) -> H::Digest {
+        let get_last_layer = |i| {
+            self.layers[TREE_DEPTH - 1]
+                .get(i)
+                .unwrap_or(&self.empty_hash_at_height[TREE_DEPTH - 1])
+        };
+        H::merge(get_last_layer(0), get_last_layer(1))
```

Comment:
I would like that! But Rust doesn't allow me to have
```rust
  layers: [Vec<H::Digest>; TREE_DEPTH + 1]
```
without enabling some unstable features. So I feel like it's cleaner to have things be `[ _  ;TREE_DEPTH]` instead of using `Vec`s instead. (Also this follows how `LocalExitTree` is defined)